components/de_amp/browser/de_amp_url_loader.cc

```diff
@@ -0,0 +1,155 @@
+
+#include "brave/components/body_sniffer/body_sniffer_url_loader.h"
+#include "brave/components/de_amp/browser/de_amp_throttle.h"
+#include "brave/components/de_amp/browser/de_amp_util.h"
+#include "mojo/public/cpp/bindings/self_owned_receiver.h"
+
+namespace de_amp {
+
+namespace {
+
+constexpr uint32_t kReadBufferSize = 65536;
+
+}  // namespace
+
+// static
+std::tuple<mojo::PendingRemote<network::mojom::URLLoader>,
+           mojo::PendingReceiver<network::mojom::URLLoaderClient>,
+           DeAmpURLLoader*>
+DeAmpURLLoader::CreateLoader(
+    base::WeakPtr<body_sniffer::BodySnifferThrottle> throttle,
+    const GURL& response_url,
+    scoped_refptr<base::SequencedTaskRunner> task_runner) {
+  mojo::PendingRemote<network::mojom::URLLoader> url_loader;
+  mojo::PendingRemote<network::mojom::URLLoaderClient> url_loader_client;
+  mojo::PendingReceiver<network::mojom::URLLoaderClient>
+      url_loader_client_receiver =
+          url_loader_client.InitWithNewPipeAndPassReceiver();
+
+  auto loader = base::WrapUnique(
+      new DeAmpURLLoader(std::move(throttle), response_url,
+                         std::move(url_loader_client), std::move(task_runner)));
+  DeAmpURLLoader* loader_rawptr = loader.get();
+  mojo::MakeSelfOwnedReceiver(std::move(loader),
+                              url_loader.InitWithNewPipeAndPassReceiver());
+  return std::make_tuple(std::move(url_loader),
+                         std::move(url_loader_client_receiver), loader_rawptr);
+}
+
+DeAmpURLLoader::DeAmpURLLoader(
+    base::WeakPtr<body_sniffer::BodySnifferThrottle> throttle,
+    const GURL& response_url,
+    mojo::PendingRemote<network::mojom::URLLoaderClient>
+        destination_url_loader_client,
+    scoped_refptr<base::SequencedTaskRunner> task_runner)
+    : body_sniffer::BodySnifferURLLoader(
+          throttle,
+          response_url,
+          std::move(destination_url_loader_client),
+          task_runner) {}
+
+DeAmpURLLoader::~DeAmpURLLoader() = default;
+
+void DeAmpURLLoader::OnBodyReadable(MojoResult) {
+  if (state_ == State::kSending) {
+    // The pipe becoming readable when kSending means all buffered body has
+    // already been sent.
+    ForwardBodyToClient();
+    return;
+  }
+  if (!CheckBufferedBody(kReadBufferSize)) {
+    return;
+  }
+
+  if (!MaybeRedirectToCanonicalLink()) {
+    CompleteLoading(std::move(buffered_body_));
+  }
+
+  body_consumer_watcher_.ArmOrNotify();
+}
+
+bool DeAmpURLLoader::MaybeRedirectToCanonicalLink() {
+  std::string canonical_link;
+
+  if (throttle_ && FindCanonicalLinkIfAMP(buffered_body_, &canonical_link)) {
+    const GURL canonical_url(canonical_link);
+    if (!VerifyCanonicalLink(canonical_url, response_url_)) {
+      VLOG(2) << __func__ << " canonical link check failed " << canonical_url;
+      return false;
+    }
+    VLOG(2) << __func__ << " de-amping and loading " << canonical_url;
+    Abort();
+    static_cast<DeAmpThrottle*>(throttle_.get())->Redirect(canonical_url);
```

Comment:
Upcast is a bad thing. Better to pass in this class the correct typed `throttle` and save it here.

Reply:
@atuchin-m that's actually not possible here because it gets passed into the superclass. This was a workaround for now.

Reply:
I think the Redirect method just needs to be on a different class like a delegate or something

Reply:
I guess for now I could just pass in the `DeAmpThrottle` to the `DeAmpURLLoader` constructor as `DeAmpThrottle`